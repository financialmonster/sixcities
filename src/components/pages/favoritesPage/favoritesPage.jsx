import React, {useEffect}  from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import {bindActionCreators} from 'redux';
import {Link, Redirect} from 'react-router-dom';
import cn from 'classnames';

import {fetchFavorites} from '../../../actions/favorites.js';
import Spinner from '../../spinner/spinner.jsx';
import Header from '../../header/header.jsx';
import FavoritesList from '../../favoritesList/favoritesList.jsx';

import logo from './logo.svg';

const FavoritesPage = ({favorites, isFavoritesLoading, isFavoritesError, isLoggedIn, fetchFavorites}) => {
    useEffect(() => {
        fetchFavorites();
    }, [fetchFavorites]);

    if(!isLoggedIn) {
        return <Redirect to="/login" />;
    }

    if(isFavoritesLoading) {
        return <Spinner />;
    }

    const divClassName = cn(`page`, {'page--favorites-empty': !favorites.length});

    return (
        <div className={divClassName}>
            <Header isMain={false} />
            <FavoritesList />
            <footer className="footer">
                <Link to="/" className="footer__logo-link" title="To the main page">
                    <img className="footer__logo" src={logo} alt="6 cities logo" width="64" height="33" />
                </Link>
            </footer>
        </div>
    );
}

FavoritesPage.propTypes = {
    favorites: PropTypes.array,
    fetchFavorites: PropTypes.func.isRequired,
    isFavoritesLoading: PropTypes.bool.isRequired,
    isFavoritesError: PropTypes.bool.isRequired,
    isLoggedIn: PropTypes.bool.isRequired
}

const mapStateToProps = ({favorites: {favorites, isFavoritesLoading, isFavoritesError}, auth: {isLoggedIn}}) => {
    return {
        favorites,
        isFavoritesLoading, 
        isFavoritesError,
        isLoggedIn
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        fetchFavorites: bindActionCreators(fetchFavorites, dispatch)
    }
}

export {FavoritesPage};
export default connect(mapStateToProps, mapDispatchToProps)(FavoritesPage);
