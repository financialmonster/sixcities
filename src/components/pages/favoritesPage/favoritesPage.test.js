import React from 'react';
import renderer from 'react-test-renderer';
import {BrowserRouter as Router} from 'react-router-dom';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';

import {FavoritesPage} from './favoritesPage.jsx';

const favorites =  [{price: 120, is_premium: true, is_favorite: false, rating: 4, src: `img/apartment-01.jpg`,
    title: `Beautiful luxurious apartment at great location`, type: `Apartment`,
    preview_image: `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/9.jpg`, city: {name: `Cologne`,
    location: {latitude: 50.938361, longitude: 6.959974, zoom: 13}}, id: 1},
    {price: 100, is_premium: false, is_favorite: true, rating: 5, src: `img/apartment-02.jpg`,
    title: `Perfectly located Castro`, type: `Room`,
    preview_image: `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/4.jpg`, city: {name: `Brussels`,
    location: {latitude: 50.846557, longitude: 4.351697, zoom: 13}}, id: 2},
    {price: 80, is_premium: true, is_favorite: true, rating: 3, src: `img/apartment-03.jpg`,
    title: `Waterfront with extraordinary view`, type: `Hotel`,
    preview_image: `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/3.jpg`,  city :{name :`Hamburg`,
    location :{latitude: 53.550341, longitude: 10.000654, zoom: 13}}, id: 3}
];

describe(`Correctly renders the FavoritesPage component`, () => {
    it(`when isLoggedIn property is equal false`, () => {
        const mockStore = configureStore([]);
        const store = mockStore();

        const tree = renderer.create(
            <Provider store={store}>
                <Router>
                    <FavoritesPage favorites={favorites} isLoggedIn={false} isFavoritesLoading={false}
                        isFavoritesError={false} fetchFavorites={() =>{}}
                    />
                </Router>
            </Provider>)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it(`when isLoggedIn and isFavoritesLoading properties are equal true and isFavoritesError property is equal
        false`,
    () => {
        const mockStore = configureStore([]);
        const store = mockStore();

        const tree = renderer.create(
            <Provider store={store}>
                <Router>
                    <FavoritesPage favorites={favorites} isLoggedIn={true} isFavoritesLoading={true}
                        isFavoritesError={false} fetchFavorites={() =>{}}
                    />
                </Router>
            </Provider>)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it(`when isLoggedIn and isFavoritesError properties are equal true and isFavoritesLoading property is equal
        false`,
    () => {
        const mockStore = configureStore([]);
        const store = mockStore();

        const tree = renderer.create(
            <Provider store={store}>
                <Router>
                    <FavoritesPage favorites={favorites} isLoggedIn={true} isFavoritesLoading={false}
                        isFavoritesError={true} fetchFavorites={() =>{}}
                    />
                </Router>
            </Provider>)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it(`when isLoggedIn property is equal true and both isFavoritesError and isFavoritesLoading properties are equal
        false`,
    () => {
        const mockStore = configureStore([]);
        const store = mockStore({auth: {isLoggedIn: false, isAuthStatusLoading: false , isAuthStatusError: false}});

        const tree = renderer.create(
            <Provider store={store}>
                <Router>
                    <FavoritesPage favorites={favorites} isLoggedIn={true} isFavoritesLoading={false}
                        isFavoritesError={false} fetchFavorites={() =>{}}
                    />
                </Router>
            </Provider>)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });
});
