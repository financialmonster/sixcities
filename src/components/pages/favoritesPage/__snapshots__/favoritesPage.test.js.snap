// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correctly renders the FavoritesPage component when isLoggedIn and isFavoritesError properties are equal true and isFavoritesLoading property is equal
        false 1`] = `
<div
  className="cities__status-wrapper tabs__content"
>
  <b
    className="cities__status"
  >
    We're sorry
  </b>
  <p
    className="cities__status-description"
  >
    It's been an error during the process of 
    loading of the list of favorite offers
    . Please, reload the page and try again
  </p>
</div>
`;

exports[`Correctly renders the FavoritesPage component when isLoggedIn and isFavoritesLoading properties are equal true and isFavoritesError property is equal
        false 1`] = `
<div
  className="spinner"
>
  <div
    className="spinnerContent"
  >
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
  </div>
</div>
`;

exports[`Correctly renders the FavoritesPage component when isLoggedIn property is equal false 1`] = `null`;

exports[`Correctly renders the FavoritesPage component when isLoggedIn property is equal true and both isFavoritesError and isFavoritesLoading properties are equal
        false 1`] = `
<div
  className="page"
>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="header__wrapper"
      >
        <div
          className="header__left"
        >
          <a
            className="header__logo-link"
            href="/"
            onClick={[Function]}
            title="To the main page"
          >
            <img
              alt="6 cities logo"
              className="header__logo"
              height="41"
              src="logo.svg"
              width="81"
            />
          </a>
        </div>
        <nav
          className="header__nav"
        >
          <ul
            className="header__nav-list"
          >
            <li
              className="header__nav-item user"
            >
              <a
                className="header__nav-link header__nav-link--profile"
                href="/login"
                onClick={[Function]}
                title="To the login page"
              >
                <div
                  className="header__avatar-wrapper user__avatar-wrapper"
                />
                <span
                  className="header__login"
                >
                  Sign in
                </span>
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
  <main
    className="page__main page__main--favorites"
  >
    <div
      className="page__favorites-container container"
    >
      <section
        className="favorites"
      >
        <h1
          className="favorites__title"
        >
          Saved listing
        </h1>
        <ul
          className="favorites__list"
        >
          <li
            className="favorites__locations-items"
          >
            <div
              className="favorites__locations locations locations--current"
            >
              <div
                className="locations__item"
              >
                <a
                  className="locations__item-link"
                  href="#"
                >
                  <span>
                    Brussels
                  </span>
                </a>
              </div>
            </div>
            <div
              className="favorites__places"
            >
              <a
                href="/offer/2"
                onClick={[Function]}
                title="To the Perfectly located Castro offer page"
              >
                <article
                  className="place-card favorites__card"
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                >
                  <div
                    className="place-card__image-wrapper favorites__image-wrapper"
                  >
                    <img
                      alt="Place pic"
                      className="place-card__image"
                      height="110"
                      src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/4.jpg"
                      width="150"
                    />
                  </div>
                  <div
                    className="place-card__info favorites__card-info"
                  >
                    <div
                      className="place-card__price-wrapper"
                    >
                      <div
                        className="place-card__price"
                      >
                        <b
                          className="place-card__price-value"
                        >
                          €
                          100
                        </b>
                        <span
                          className="place-card__price-text"
                        >
                          / night
                        </span>
                      </div>
                      <button
                        className="place-card__bookmark-button button place-card__bookmark-button--active"
                        title="In bookmarks"
                        type="button"
                      >
                        <svg
                          className="place-card__bookmark-icon"
                          height="19"
                          width="18"
                        >
                          <use
                            xlinkHref="#icon-bookmark"
                          />
                        </svg>
                        <span
                          className="visually-hidden"
                        >
                          In bookmarks
                        </span>
                      </button>
                    </div>
                    <div
                      className="place-card__rating rating"
                    >
                      <div
                        className="place-card__stars rating__stars"
                      >
                        <span
                          style={
                            Object {
                              "width": "100%",
                            }
                          }
                        />
                        <span
                          className="visually-hidden"
                        >
                          Rating
                        </span>
                      </div>
                    </div>
                    <h2
                      className="place-card__name"
                    >
                      Perfectly located Castro
                    </h2>
                    <p
                      className="place-card__type"
                    >
                      Room
                    </p>
                  </div>
                </article>
              </a>
            </div>
          </li>
          <li
            className="favorites__locations-items"
          >
            <div
              className="favorites__locations locations locations--current"
            >
              <div
                className="locations__item"
              >
                <a
                  className="locations__item-link"
                  href="#"
                >
                  <span>
                    Cologne
                  </span>
                </a>
              </div>
            </div>
            <div
              className="favorites__places"
            >
              <a
                href="/offer/1"
                onClick={[Function]}
                title="To the Beautiful luxurious apartment at great location offer page"
              >
                <article
                  className="place-card favorites__card"
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                >
                  <div
                    className="place-card__mark"
                  >
                    <span>
                      Premium
                    </span>
                  </div>
                  <div
                    className="place-card__image-wrapper favorites__image-wrapper"
                  >
                    <img
                      alt="Place pic"
                      className="place-card__image"
                      height="110"
                      src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/9.jpg"
                      width="150"
                    />
                  </div>
                  <div
                    className="place-card__info favorites__card-info"
                  >
                    <div
                      className="place-card__price-wrapper"
                    >
                      <div
                        className="place-card__price"
                      >
                        <b
                          className="place-card__price-value"
                        >
                          €
                          120
                        </b>
                        <span
                          className="place-card__price-text"
                        >
                          / night
                        </span>
                      </div>
                      <button
                        className="place-card__bookmark-button button"
                        title="To bookmarks"
                        type="button"
                      >
                        <svg
                          className="place-card__bookmark-icon"
                          height="19"
                          width="18"
                        >
                          <use
                            xlinkHref="#icon-bookmark"
                          />
                        </svg>
                        <span
                          className="visually-hidden"
                        >
                          Not in bookmarks
                        </span>
                      </button>
                    </div>
                    <div
                      className="place-card__rating rating"
                    >
                      <div
                        className="place-card__stars rating__stars"
                      >
                        <span
                          style={
                            Object {
                              "width": "80%",
                            }
                          }
                        />
                        <span
                          className="visually-hidden"
                        >
                          Rating
                        </span>
                      </div>
                    </div>
                    <h2
                      className="place-card__name"
                    >
                      Beautiful luxurious apartment at great location
                    </h2>
                    <p
                      className="place-card__type"
                    >
                      Apartment
                    </p>
                  </div>
                </article>
              </a>
            </div>
          </li>
          <li
            className="favorites__locations-items"
          >
            <div
              className="favorites__locations locations locations--current"
            >
              <div
                className="locations__item"
              >
                <a
                  className="locations__item-link"
                  href="#"
                >
                  <span>
                    Hamburg
                  </span>
                </a>
              </div>
            </div>
            <div
              className="favorites__places"
            >
              <a
                href="/offer/3"
                onClick={[Function]}
                title="To the Waterfront with extraordinary view offer page"
              >
                <article
                  className="place-card favorites__card"
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                >
                  <div
                    className="place-card__mark"
                  >
                    <span>
                      Premium
                    </span>
                  </div>
                  <div
                    className="place-card__image-wrapper favorites__image-wrapper"
                  >
                    <img
                      alt="Place pic"
                      className="place-card__image"
                      height="110"
                      src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/3.jpg"
                      width="150"
                    />
                  </div>
                  <div
                    className="place-card__info favorites__card-info"
                  >
                    <div
                      className="place-card__price-wrapper"
                    >
                      <div
                        className="place-card__price"
                      >
                        <b
                          className="place-card__price-value"
                        >
                          €
                          80
                        </b>
                        <span
                          className="place-card__price-text"
                        >
                          / night
                        </span>
                      </div>
                      <button
                        className="place-card__bookmark-button button place-card__bookmark-button--active"
                        title="In bookmarks"
                        type="button"
                      >
                        <svg
                          className="place-card__bookmark-icon"
                          height="19"
                          width="18"
                        >
                          <use
                            xlinkHref="#icon-bookmark"
                          />
                        </svg>
                        <span
                          className="visually-hidden"
                        >
                          In bookmarks
                        </span>
                      </button>
                    </div>
                    <div
                      className="place-card__rating rating"
                    >
                      <div
                        className="place-card__stars rating__stars"
                      >
                        <span
                          style={
                            Object {
                              "width": "60%",
                            }
                          }
                        />
                        <span
                          className="visually-hidden"
                        >
                          Rating
                        </span>
                      </div>
                    </div>
                    <h2
                      className="place-card__name"
                    >
                      Waterfront with extraordinary view
                    </h2>
                    <p
                      className="place-card__type"
                    >
                      Hotel
                    </p>
                  </div>
                </article>
              </a>
            </div>
          </li>
        </ul>
      </section>
    </div>
  </main>
  <footer
    className="footer"
  >
    <a
      className="footer__logo-link"
      href="/"
      onClick={[Function]}
      title="To the main page"
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="logo.svg"
        width="64"
      />
    </a>
  </footer>
</div>
`;
